Move to GenomicsDB 1.2.0

nalinigans opened the pull request:
Move to GenomicsDB 1.2.0. This version uses a 64-bit [htslib](https://github.com/GenomicsDB/htslib) to workaround overflow issues when computed annotation sizes exceed the 32-bit integer space as requested by @ldgauthier.

The htslib and GenomicsDB fixes by @kgururaj also enable read/write of VCFs where the values of INFO fields can be 64-bits. But, if the input VCF (produced by HaplotypeCaller) contains INFO fields with values larger than what can fit in 32 bits, the vid must list those fields as type int64 (currently, it will be int by default). However, no changes are needed in the vid if the input values fit within 32 bits but the result of the combine operation [e.g. sum] is larger.

Note that 64-bit integers are currently only supported for INFO fields (no support for FORMAT fields).

kgururaj commented:
I believe this release also fixes #6158 and #6275 (thanks @mlathara).

nalinigans commented:
GenomicsDB 1.2.0 also supports suppression of "No valid combination operation" warnings. This will however require changes in gatk(maybe hardcoded initially via GenomicsDBUtils) to indicate to GenomicsDB that a set of fields have no combinations. I will open another PR with an example.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java:
```diff
@@ -396,7 +397,7 @@ final void printCacheStats() {
                     createExportConfiguration(workspace, callsetJson, vidmapJson, vcfHeader, genomicsDBOptions);
             return new GenomicsDBFeatureReader<>(exportConfigurationBuilder,
-                    new GenomicsDBBCFCodec(),
+                    new VCFCodec(),
```
Wondering about the performance implications of changing to VCF format ? Also, if we no longer need the  `GenomicsDBBCFCodec` codec, that class (and it's tests) can be deleted.

cmnbroad replied:
Also, there are some other tests where `GenomicsDBBCFCodec` is still used - I'm surprised those don't seem to be failing.

nalinigans replied:
GenomicsDBImportIntegrationTest has been ported already to use VCFCodec in the change set for this PR. But, I can parameterize the tests to test both BCFCodec and VCFCodec if decide to keep BCFCodec for performance reasons.

Regarding performance, we will have to benchmark GenomicsDB queries as @droazen mentions. We can probably add a FeatureReader option to choose between BCFCodec and VCFCodec for GenomicsDB if performance turns out to be an issue. If BCFCodec is chosen, GenomicsDB gracefully handles a number of overflow scenarios already.

Note that -
* BCF and BCFCodec do not support 64-bit fields.
* VCFCodec does not fully decode the VCF streams by default, it uses the LazyParser whereas BCFCodec always fully decodes the BCF streams. Hopefully, this keeps the performance somewhat similar even if fields have to be explicitly decoded.
* GenomicsDBImport already uses VCFCodec to read vcfs, so no change there.

cmnbroad replied:
Ah, ok, I see. So then if we just need to delete the `GenomicsDBBCFCodec` class (unless we keep an option for BCF, in which case it will depend on the version of BCF that uses). The `GenomicsDBBCFCodec` is only necessary if the BCF header is stamped with version 2.2.

kgururaj replied:
I did a benchmark run with about 5K samples more than a year back - negligible difference in performance between VCF and BCF serialization/deseralization. It would be good to re-check again with your data.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureDataSource.java:
```diff
@@ -396,7 +397,7 @@ final void printCacheStats() {
                     createExportConfiguration(workspace, callsetJson, vidmapJson, vcfHeader, genomicsDBOptions);
             return new GenomicsDBFeatureReader<>(exportConfigurationBuilder,
-                    new GenomicsDBBCFCodec(),
+                    new VCFCodec(),
```
As discussed in person, we should benchmark this branch to make sure that the switch to `VCFCodec` did not cause a major performance regression.

nalinigans commented:
@ldgauthier, does this branch fix the overflow with combined annotation for RAW_MQandDP in the GenomicsDB you encountered for the large cohort? Thanks.

ldgauthier commented:
@nalinigans I can test on my use case specifically, but it's going to take me some time to dig up all the data again.

nalinigans commented:
@droazen, I have introduced --genomicsdb-use-vcf-codec argument defaulting to false. This will allow the user to choose between VCFCodec and BCF2Codec for GenomicsDB streams.

```
--genomicsdb-use-vcf-codec:Boolean
                              Use VCF Codec Streaming for data from GenomicsDB  Default value: true. Possible values:
                              {true, false} 
```

Also, hooked --max-genotype-count arg from GenotypeGVCFs/GnarlyGenotyper to GenomicsDB Export Configuration and introduced --genomicsdb-max-genotype-count to be used by tools like SelectVariants that do not use the Genotype ArgumentCollection.

```
--genomicsdb-max-genotype-count:Integer
                              Maximum number of genotypes to consider at any site  Default value: 1024.
```

nalinigans commented:
@ldgauthier, you mentioned that it would take some time for you to dig into this. But, does this branch fix the overflow with combined annotation for RAW_MQandDP in the GenomicsDB you encountered for the large cohort? Please note that you have to explicitly set the useVCFCodec arg/option to true to test.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBOptions.java:
```diff
@@ -10,26 +10,32 @@
+        this.callGenotypes = genomicsdbArgs.callGenotypes;
+        this.maxDiploidAltAllelesThatCanBeGenotyped = genomicsdbArgs.maxDiploidAltAllelesThatCanBeGenotyped;
+        this.maxGenotypeCount = genotypeCalcArgs.MAX_GENOTYPE_COUNT;
+        this.useVCFCodec = genomicsdbArgs.useVCFCodec;
```
@nalinigans This change can be deferred and made in a subsequent PR if you'd prefer.

nalinigans replied:
Yes, let's defer this to another PR. Thanks.

nalinigans commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java:
```diff
@@ -56,6 +58,22 @@
+     * @param v2 List of expected variant contexts
+     */
+    private static void assertEqualVariants(
+            final List<VariantContext> v1, final List<VariantContext> v2) {
```
Done.

nalinigans commented on src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java:
```diff
@@ -713,6 +712,16 @@ private static File getSampleMapFile(final Map<String, String> mapping){
+     * Genotype.getAnyAttribute() returns String or Integer based on the VC being decoded fully, so use this routine instead to
+     * guarantee that it is an int.
+     */
+    private static int getAttributeAsInt(Genotype genotype, String key) {
```
Done.

nalinigans commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBBaseArgumentCollection.java:
```diff
@@ -0,0 +1,34 @@
+      fullName = USE_VCF_CODEC_LONG_NAME,
+      doc = "Use VCF Codec Streaming for data from GenomicsDB",
+      optional = true)
+  public boolean useVCFCodec = DEFAULT_USE_VCF_CODEC;
```
Actually, we need the default be BCF as we are not sure about the performance implications of VCF as yet. Changed the `DEFAULT_USE_VCF_CODEC` to false

nalinigans replied:
The intention was to have the default as false.

nalinigans replied:
Have changed the GenomicsDB/GnarlyGenotyper tests to run with both codecs. But, I don't have the bandwidth to introduce new 64-bit VCFs to test VCF Codec. We do have tests in GenomicsDB for the 64 bit VCFs - see [PR](https://github.com/GenomicsDB/GenomicsDB/pull/73) and vcfs like [bigint0.vcf.gz](https://github.com/GenomicsDB/GenomicsDB/blob/master/tests/inputs/vcfs/bigint0.vcf.gz) for the GenomicsDB CI tests and they can be ported to work with GATK test references. I can look into this later or perhaps it is easy enough for someone else from the team to do it.

nalinigans commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBArgumentCollection.java:
```diff
@@ -0,0 +1,18 @@
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.hellbender.tools.walkers.genotyper.GenotypeCalculationArgumentCollection;
+
+import java.io.Serializable;
```
Fixed this.

nalinigans commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBArgumentCollection.java:
```diff
@@ -0,0 +1,18 @@
+package org.broadinstitute.hellbender.tools.genomicsdb;
+
+import htsjdk.variant.variantcontext.GenotypeLikelihoods;
```
Fixed this.

nalinigans commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java:
```diff
@@ -56,6 +58,22 @@
+        Utils.nonNull(v2, "v2");
+        Assert.assertEquals(v1.size(), v2.size());
+        for (int i = 0; i < v1.size(); i++) {
+            VariantContextTestUtils.assertVariantContextsAreEqual(v1.get(i), v2.get(i), new ArrayList<String>(), new ArrayList<String>());
```
Fixed using Collections.emptyList(). Thanks.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java:
```diff
@@ -420,6 +423,9 @@
     private boolean suppressReferencePath = false;
 
+    @ArgumentCollection
+    private GenomicsDBArgumentCollection genomicsdbArgs = new GenomicsDBArgumentCollection();
```
Has any thought been given to how this new argument interacts with the existing genotype subsetting arguments SelectVariants already has ? If not, I'm thinking this change will likely cause confusion, and that we should take this out until those are resolved and some tests are added.

nalinigans replied:
There is no existing argument that clashes with the existing subsetting arguments. However, if you would rather take this out, please let me know.

cmnbroad replied:
SV has several arguments that do various kinds of genotype subsetting/downsampling/filtering. The argument is marked as `@Advanced`, so maybe its ok. But if this argument does what I think it does, it may not be obvious to a user that this subsetting is done independently before SV ever sees the data.

nalinigans replied:
This needs more thought. Have removed all support for GenomicsDB args in SelectVariants for now and have filed #6456 to generally get the tools hook up to GenomicsDBExportConfiguration better.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/gnarlyGenotyper/GnarlyGenotyper.java:
```diff
@@ -169,7 +172,9 @@ public boolean requiresReference() {
         if (genomicsDBOptions == null) {
-            genomicsDBOptions = new GenomicsDBOptions(referenceArguments.getReferencePath(), CALL_GENOTYPES, PIPELINE_MAX_ALT_COUNT);
+            genomicsdbArgs.callGenotypes = CALL_GENOTYPES;
+            genomicsdbArgs.maxDiploidAltAllelesThatCanBeGenotyped = PIPELINE_MAX_ALT_COUNT;
```
Its a bit strange to mutate values in an argument collection like this (I know they aren't annotated as `@Argument`, but presumably someday they will be exposed ?). Can you verify that the only net change here in this part of the PR is that we're exposing the vcfCodec switch to `GnarlyGenotyper` (?).

nalinigans replied:
Yes, the net change here allows GnarlyGenotyper to access the vcfCodec arg. But, have also hooked `--max-genotype-count` arg from GenotypeGVCFs/GnarlyGenotyper to be passed to GenomicsDBOptions to help with Issue #6275.

nalinigans commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/gnarlyGenotyper/GnarlyGenotyperEngine.java:
```diff
@@ -584,4 +576,34 @@ private static String trimASAnnotation(final VariantContext variant, final List<
+   * @param g genotype for a sample
+   * @return int array for the given genotype
+   */
+  private static int[] getSBFieldAsIntArray(Genotype g) {
```
I just refactored the existing code with existing patterns of throwing exceptions in GnarlyGenotyper. I will leave to @ldgauthier to clean this up further. Thanks.

nalinigans commented:
I have merged to the latest master, but Travis builds are still failing not related to the changes in this PR. @droazen, @cmnbroad, can someone take a look and let me know what should be done. Thanks.

cmnbroad commented:
@nalinigans thanks for the updates. All builds are failing for everyone at the moment, so its not just this branch. We're working on a fix. I'll take a look at your changes.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBArgumentCollection.java:
```diff
@@ -0,0 +1,15 @@
+  private static final String MAX_GENOTYPE_COUNT_LONG_NAME = "genomicsdb-max-genotype-count";
+
+  @Advanced
+  @Argument(fullName=MAX_GENOTYPE_COUNT_LONG_NAME, doc="Maximum number of genotypes to consider at any site", optional=true)
```
Since this is only used outside outside of the genotyping tools, I think it would make sense to remove the word "consider" from the doc string, and also reinforce that this only applies when the underlying store is GenomicsDB. Maybe something like "Maximum number of genotypes to retrieve when the underlying store is GenomicsDB".

nalinigans replied:
This argument was being used only by SelectVariants. Note that the existing `GenomicsDBArgumentCollection` is removed and `GenomicsDBBaseArgumentCollection` is renamed to `GenomicsDBArgumentCollection`.

cmnbroad commented on src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java:
```diff
@@ -520,6 +522,33 @@ private static void checkGenomicsDBAgainstExpected(final String workspace, final
+                false);
+    }
+
+    private static void checkGenomicsDBAgainstExpected(final String workspace, final List<SimpleInterval> intervals,
```
This appears to be unused.

nalinigans replied:
Done
